state/upgrades.go

```diff
@@ -481,6 +482,80 @@ func updateLegacyLXDCredentialsOps(st *State, cred cloud.Credential) ([]txn.Op,
 	return ops, nil
 }
 
+// UpdateLegacyKubernetesCloudCredentials updates the cloud credentials for
+// Kubernetes clouds. This is to introduce bug fix changes from 2.9 onwards.
+func UpdateLegacyKubernetesCloudCredentials(st *State) error {
+	ops, err := updateLegacyKubernetesCloudAuthTypesOps(st)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return st.db().RunTransaction(ops)
+}
+
+func updateLegacyKubernetesCloudAuthTypesOps(st *State) ([]txn.Op, error) {
+	clouds, err := st.Clouds()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var ops []txn.Op
+	for _, c := range clouds {
+		if c.Type != "kubernetes" {
+			continue
+		}
+		set := bson.D{{"auth-types", k8scloud.SupportedAuthTypes()}}
+		ops = append(ops, txn.Op{
+			C:      cloudsC,
+			Id:     c.Name,
+			Assert: txn.DocExists,
+			Update: bson.D{{"$set", set}},
+		})
+
+		credOps, err := updateLegacyKubernetesCredentialsOps(st, c.Name)
+		if err != nil {
+			return nil, errors.Annotatef(err, "updating legacy kubernetes credentials for cloud %s", c.Name)
+		}
+		ops = append(ops, credOps...)
+	}
+	return ops, nil
+}
+
+func updateLegacyKubernetesCredentialsOps(st *State, cloudName string) ([]txn.Op, error) {
+	coll, closer := st.db().GetRawCollection(cloudCredentialsC)
+	defer closer()
+	iter := coll.Find(bson.M{"cloud": cloudName}).Iter()
+	defer iter.Close()
+
+	var ops []txn.Op
+	var doc cloudCredentialDoc
```

Comment:
In upgrade steps, we avoid using the entity doc - we read / write bson maps instead as these are generic abd not affected by any incompatible changes as the code evolved

Reply:
Done & Done